src/envoy/http/error_translator/filter.cc

```diff
@@ -0,0 +1,211 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/envoy/http/error_translator/filter.h"
+
+#include <string>
+
+#include "absl/types/optional.h"
+#include "common/buffer/buffer_impl.h"
+#include "common/grpc/status.h"
+#include "common/http/headers.h"
+#include "src/envoy/utils/filter_state_utils.h"
+
+namespace espv2 {
+namespace envoy {
+namespace http_filters {
+namespace error_translator {
+
+using ::Envoy::Http::FilterDataStatus;
+using ::Envoy::Http::FilterHeadersStatus;
+
+Filter::Filter(FilterConfigSharedPtr config) : config_(config) {}
+
+FilterHeadersStatus Filter::encodeHeaders(
+    Envoy::Http::ResponseHeaderMap& headers, bool) {
+  if (isUpstreamResponse()) {
+    // Do not translate any response sent from the upstream.
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (headers.ContentType() != nullptr &&
+      headers.ContentType()->value().getStringView() ==
+          Envoy::Http::Headers::get().ContentTypeValues.Grpc) {
+    is_grpc_response_ = true;
+  }
+
+  if (isEspv2FilterError() && is_grpc_response_) {
+    // Do not translate for gRPC, body does not matter.
+    // gRPC error is already in filter state as well, so nothing to do.
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (isEspv2FilterError()) {
+    // We need to modify the response body, but keep filter state as is.
+    ENVOY_LOG(debug, "Translating HTTP espv2 error headers");
+
+    // Store a copy and scrub details if needed.
+    error_ = utils::getErrorFilterState(
+        *encoder_callbacks_->streamInfo().filterState());
+    if (config_->shouldScrubDebugDetails()) {
+      error_.clear_details();
+    }
+
+    // Translate to JSON.
+    error_json_ = errorToJson(error_);
+
+    // Update content headers.
+    headers.setContentLength(error_json_.size());
+    headers.setReferenceContentType(
+        Envoy::Http::Headers::get().ContentTypeValues.Json);
+
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (is_grpc_response_) {
+    // We keep the response body as is, but modify the filter state.
+    // This will be used by the access log (SC Report).
+    // This handles the trailers-only response by `sendLocalReply`.
```

Comment:
Is this error from our ESPv2 filters,  if so,  we already have rpc.Status in the filterState.  Do we need to do this?

Reply:
There is one case that we can write rpc.Status to our filterState:

non_espv2  && non_grpc:

the error message is in the response body,  we need to convert it from plain text to JSON,

In order to send this error message in the Report, we can write it to stateFilter too.

Reply:
Good point on it adding latency. Can we put it here for now and move it out in a different PR? Otherwise, I'm worried about the added complexity. Anyways, this code is not used yet.

Re comment #2 above, yup. I plan to keep this case writing the filter state, this is the end of `encodeData`.